File: java-junit/src/test/java/org/codecop/Session5b_WordCounterRatioTest.java
package org.codecop;

import static org.junit.Assert.assertEquals;
import junitparams.JUnitParamsRunner;
import junitparams.Parameters;

import org.junit.Test;
import org.junit.runner.RunWith;

/**
 * Session 5b: WordCounterRatioTest - parameterised/table driven tests with JUnitParams. <br />
 * @see "https://github.com/Pragmatists/junitparams"
 */
@RunWith(JUnitParamsRunner.class) // Mark this test as parameterised with JUnitParams
public class Session5b_WordCounterRatioTest {

    // TODO Add the needed annotations to run this test with all examples.
    
    @Test // keep
    // We want to test more cases for the ratio. Here is a table of test cases.
    @Parameters(// // drop
        // use /*
        {
        "green, green, 1.0", //
        "green bar green, green, 0.66", //
        "green bar green bar, green, 0.5", //
        "green bar green, bar, 0.33" //
        }
        // use */
    ) // drop
    public void shouldReturnRatioOfGivenWord(String sentence, String word, double expectedRatio) {
        WordCounter counter = new WordCounter(sentence);
        assertEquals(expectedRatio, counter.ratioOf(word), 0.01); // keep
    }

}

File: csharp-xunit/WordCountTests/Session5b_WordCounterRatioTest.cs
using System.Collections;
using System.Collections.Generic;
using Org.Codecop.WordCount;
using Xunit;

namespace Org.Codecop.WordCount.Tests
{
    /// <summary>Session 5b: WordCounterRatioTest - parameterised/table driven tests with ClassData.</summary>
    /// <seealso>"http://hamidmosalla.com/2017/02/25/xunit-theory-working-with-inlinedata-memberdata-classdata/"</seealso>
    public class Session5b_WordCounterRatioTest
    {

        // TODO Add the needed attributes to run this test with all examples.
        [Theory] // Mark this test as parameterised.
        [ClassData(typeof(TestCases))] // Tell it to take arguments from "TestCases".
        public void ShouldReturnRatioOfGivenWord(string sentence, string word, double expectedRatio)
        {
            var counter = new WordCounter(sentence);
            Assert.Equal(expectedRatio, counter.RatioOf(word), 2); // keep
        }

    }

    public class TestCases : IEnumerable<object[]>
    {
        // We want to test more corner cases for the ratio. Here is a table of test cases.
        private readonly List<object[]> testCases = new List<object[]>{
            new object[] {"green", "green", 1.0},
            new object[] {"green bar green", "green", 0.67},
            new object[] {"green bar green bar", "green", 0.5},
            new object[] {"green bar green", "bar", 0.33}
        };

        public IEnumerator<object[]> GetEnumerator() => testCases.GetEnumerator();

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}
